src/main/java/net/neoforged/neoforge/transfer/resources/FluidResource.java

```diff
@@ -0,0 +1,336 @@
+    final FluidStack innerStack;
+
+    /**
+     * Lazily initialized.
+     */
+    @Nullable
+    private ItemResource filledBucket;
+
+    private FluidResource(FluidStack innerStack) {
+        this.innerStack = innerStack;
+    }
+
+    /**
+     * @return The {@link Fluid} of this resource from the inner {@link FluidStack}
+     */
+    @Override
+    public Fluid getInstanceValue() {
+        return innerStack.getFluid();
+    }
+
+    /**
+     * @return the fluid holder of this resource
+     */
+    @Override
+    public Holder<Fluid> getHolder() {
+        return innerStack.getFluidHolder();
+    }
+
+    /**
+     * Checks if this resource is empty. The resource will be empty if the fluid is {@link Fluids#EMPTY}.
+     *
+     * @return if this resource is empty
+     */
+    @Override
+    public boolean isEmpty() {
+        return innerStack.isEmpty();
+    }
+
+    /**
+     * Returns a copy of this resource with the patch applied.
+     *
+     * @param patch the patch to apply
+     * @return the new resource
+     */
+    @Override
+    public FluidResource withPatch(DataComponentPatch patch) {
+        FluidStack stack = innerStack.copy();
+        stack.applyComponents(patch);
+        return new FluidResource(stack);
+    }
+
+    /**
+     * Returns a copy of this resource with the set data component.
+     *
+     * @param type the type of data component
+     * @param data the data to set
+     * @param <D>  the type of data component
+     * @return the new resource
+     */
+    @Override
+    public <D> FluidResource with(DataComponentType<D> type, D data) {
+        FluidStack stack = innerStack.copy();
+        stack.set(type, data);
+        return new FluidResource(stack);
+    }
+
+    /**
+     * Returns a copy of this resource with the data component removed.
+     *
+     * @param type The type of data component
+     * @return The new resource
+     */
+    @Override
+    public FluidResource without(DataComponentType<?> type) {
+        FluidStack stack = innerStack.copy();
+        stack.remove(type);
+        return new FluidResource(stack);
+    }
+
+    /**
+     * @return the fluid type of this resource
+     */
+    public FluidType getFluidType() {
+        return innerStack.getFluidType();
+    }
+
+    @Override
+    public DataComponentMap getComponents() {
+        if (innerStack.isEmpty()) return DataComponentMap.EMPTY;
+        return innerStack.getComponents().toImmutableMap();
+    }
+
+    @Override
+    public DataComponentPatch getComponentsPatch() {
+        return innerStack.getComponentsPatch();
+    }
+
+    public FluidStack toStack(int amount) {
+        return amount == 0 || this.isEmpty() ? FluidStack.EMPTY : this.innerStack.copyWithAmount(amount);
+    }
+
+    /**
+     * @return A {@link FluidStack} copy of the inner stack. The size is by default {@value FluidType#BUCKET_VOLUME}
+     */
+    public FluidStack toStack() {
```

Comment:
Remove this method and make the user always have to specify an amount in `toStack`.

Reply:
I liked this method, untill I asked myself how much the default value would be.
Yeah this would indeed be confusing. Please remove or provide `toBucketStack()` and `toEmptyStack()` helpers. Your choice.